File: server/src/collectors/jpNewsRss.js
/**
 * Aggregated JP news / government RSS bundle (English + Japanese).
 *
 * Free, no auth. Pulls multiple feeds in parallel; each article becomes an
 * intel item tagged with its publisher.
 */

import { intelEnvelope, feedItemToIntel, intelHashKey, intelUid } from '../utils/intelHelpers.js';
import { fetchFeed } from '../utils/rssCollectorTemplate.js';

const SOURCE_ID = 'jp-news-rss';

const FEEDS = [
  { name: 'Nikkei Asia',        url: 'https://asia.nikkei.com/rss/feed/nar',        lang: 'en' },
  { name: 'Japan Times',        url: 'https://www.japantimes.co.jp/feed/',          lang: 'en' },
  { name: 'NHK World (top)',    url: 'https://www3.nhk.or.jp/rss/news/cat0.xml',    lang: 'ja' },
  { name: 'Reuters Japan',      url: 'https://jp.reuters.com/rss',                  lang: 'ja' },
  { name: 'Kantei (PMO)',       url: 'https://japan.kantei.go.jp/rss.xml',          lang: 'en' },
  { name: 'NDL Press Releases', url: 'https://www.ndl.go.jp/en/news/index.rss',     lang: 'en' },
];
const TIMEOUT_MS = 12000;
const PER_FEED_LIMIT = 20;

export default async function collectJpNewsRss() {
  const lists = await Promise.all(FEEDS.map((f) => fetchFeed(f.url, { timeoutMs: TIMEOUT_MS })));
  const items = [];
  lists.forEach((entries, fi) => {
    const feed = FEEDS[fi];
    entries.slice(0, PER_FEED_LIMIT).forEach((e) => {
      const it = feedItemToIntel(SOURCE_ID, e, {
        language: feed.lang,
        tags: ['news', `publisher:${feed.name}`],
      });
      // Make uid include publisher so the same headline reposted across feeds
      // doesn't collide.
      it.uid = intelUid(SOURCE_ID, e.guid, e.link, intelHashKey(feed.name, e.title, e.pubDate));
      it.properties = { ...(it.properties || {}), publisher: feed.name };
      items.push(it);
    });
  });

  return intelEnvelope({
    sourceId: SOURCE_ID,
    items,
    description: 'Aggregated JP news / government RSS (Nikkei, Japan Times, Reuters JP, NHK World, Kantei, NDL)',
    extraMeta: { publishers: FEEDS.map((f) => f.name) },
  });
}

File: native/collectors/sources/jp_news_rss.c
/* collectors/social/sources/jp_news_rss.c — port of server/src/collectors/jpNewsRss.js
 * Aggregated JP news / government RSS bundle. JS loops FEEDS calling
 * fetchFeed+feedItemToIntel per feed with per-feed language + tags
 * ['news','publisher:<name>']. Ported faithfully as a per-feed rss_collect
 * loop; each feed carries its own publisher tag. The intel-envelope/_meta and
 * the publisher-hashed uid refinement (intelHashKey) are not portable through
 * the shared rss_collect helper (no uid-override hook) — uid precedence
 * guid→link→sha1(title|pubDate) mirrors intelUid for the common case. */
#include "source.h"
#include "lib/rss_atom.h"
#include <stdio.h>

struct feed { const char *url; const char *lang; const char *tags; };

static int run(const source_ctx *ctx, intel_sink *sink) {
  static const struct feed FEEDS[] = {
    { "https://asia.nikkei.com/rss/feed/nar",     "en", "[\"news\",\"publisher:Nikkei Asia\"]" },
    { "https://www.japantimes.co.jp/feed/",       "en", "[\"news\",\"publisher:Japan Times\"]" },
    { "https://www3.nhk.or.jp/rss/news/cat0.xml", "ja", "[\"news\",\"publisher:NHK World (top)\"]" },
    { "https://jp.reuters.com/rss",               "ja", "[\"news\",\"publisher:Reuters Japan\"]" },
    { "https://japan.kantei.go.jp/rss.xml",       "en", "[\"news\",\"publisher:Kantei (PMO)\"]" },
    { "https://www.ndl.go.jp/en/news/index.rss",  "en", "[\"news\",\"publisher:NDL Press Releases\"]" },
  };
  int total = 0, ok = 0;
  for (unsigned i = 0; i < sizeof(FEEDS) / sizeof(FEEDS[0]); i++) {
    int n = rss_collect(ctx, sink, FEEDS[i].url, FEEDS[i].lang, FEEDS[i].tags);
    if (n >= 0) { total += n; ok = 1; }
  }
  fprintf(stderr, "[jp-news-rss] emitted %d\n", total);
  return ok ? 0 : -1;
}

static const source_def jp_news_rss_def = {
  .id = "jp-news-rss", .collector = "social",
  .name = "JP news RSS bundle", .name_ja = "日本ニュース RSS バンドル",
   .update_interval_sec = 1800, .run = run };
REGISTER_SOURCE(jp_news_rss_def)
